examples/experimental/bert_squad_pytorch/model_def.py

```diff
@@ -0,0 +1,166 @@
+
+    def build_validation_data_loader(self) -> DataLoader:
+        return DataLoader(
+            self.validation_dataset,
+            batch_size=self.context.get_per_slot_batch_size(),
+        )
+
+    def build_model(self) -> torch.nn.modules.module.Module:
+        cache_dir_per_rank = f"/tmp/{self.context.distributed.get_rank()}"
+
+        config = self.config_class.from_pretrained(
+            self.context.get_data_config().get("model_name_or_path"),
+            cache_dir=cache_dir_per_rank,
+        )
+        model = self.model_class.from_pretrained(
+            self.context.get_data_config().get("model_name_or_path"),
+            from_tf=bool(".ckpt" in self.context.get_data_config().get("model_name_or_path")),
+            config=config,
+            cache_dir=cache_dir_per_rank,
+        )
+        return model
+
+    def optimizer(self, model: nn.Module) -> torch.optim.optimizer.Optimizer:
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [
+                    p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.context.get_hparam("weight_decay"),
+            },
+            {
+                "params": [
+                    p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.context.get_hparam("weight_decay"),
+            },
+        ]
+        optimizer = AdamW(
+            optimizer_grouped_parameters,
+            lr=self.context.get_hparam("learning_rate"),
+            eps=self.context.get_hparam("adam_epsilon")
+        )
+        return optimizer
+
+    def create_lr_scheduler(self, optimizer: torch.optim.Optimizer) -> LRScheduler:
+        scheduler = get_linear_schedule_with_warmup(
+            optimizer,
+            num_warmup_steps=self.context.get_hparam("num_warmup_steps"),
+            num_training_steps=self.context.get_hparam("num_training_steps"),
+        )
+        return LRScheduler(scheduler, LRScheduler.StepMode.STEP_EVERY_BATCH)
+
+    def train_batch(self, batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int) -> Union[torch.Tensor, Dict[str, Any]] :
+        inputs = {
+            "input_ids": batch[0],
+            "attention_mask": batch[1],
+            "token_type_ids": batch[2],
+            "start_positions": batch[3],
+            "end_positions": batch[4],
+        }
+        outputs = model(**inputs)
+        loss = outputs[0]
+        self.tr_loss += loss.item()
+        return {"loss": self.tr_loss}
+
+    def evaluate_full_dataset(self, data_loader: DataLoader, model: nn.Module) -> Dict[str, Any]:
+        tokenizer = self.tokenizer_class.from_pretrained(
+            self.context.get_data_config().get("model_name_or_path"),
+            do_lower_case=True,
+            cache_dir=None
+        )
```

Comment:
blocking: since you can reuse the tokenizer across multiple validation steps, you could move it to the trial instance constructor as `self.tokenizer`.

Reply:
actually, it looks like the same tokenizer is loaded in `load_and_cache_examples`, could you merge these paths to use the same tokenizer object? Probably only need to load it once at the beginning of the constructor.